app/src/pages/playground/ModelComboBox.tsx

```diff
@@ -1,92 +1,131 @@
-          model.name.toLocaleLowerCase() === modelInput.toLocaleLowerCase()
-      )
-    ) {
-      items.push({ name: modelInput, id: modelInput });
-    }
-    if (
-      props.modelName &&
-      !items.some((model) => model.name === props.modelName)
-    ) {
-      items.push({ name: props.modelName, id: props.modelName });
+  const [fieldState, setFieldState] = useState({
+    selectedKey: modelName ?? null,
+    inputValue: modelName ?? "",
+  });
+
+  const onSelectionChange = (key: Key | null) => {
+    if (typeof key === "string") {
+      const item = items.find((item) => item.id === key);
+      item?.name != null && onChange(item.name);
     }
-    return items;
-  }, [data.models, modelInput, props.modelName]);
+  };
+
+  const onInputChange = (value: string) => {
+    setFieldState((prevState) => ({
+      inputValue: value,
+      selectedKey: prevState.selectedKey,
+    }));
+  };
+
+  useEffect(() => {
+    setFieldState({
+      selectedKey: modelName ?? null,
+      inputValue: modelName ?? "",
+    });
+  }, [modelName]);
+
   return (
     <ComboBox
-      label={"Model"}
       size="L"
       data-testid="model-picker"
-      // Fallback to empty string here otherwise the picker will complain about switching from a controlled to uncontrolled component
-      // It can't distinguish between undefined and intentionally null
-      selectedKey={props.modelName ?? ""}
+      selectedKey={fieldState.selectedKey}
       aria-label="model picker"
-      onInputChange={setModelInput}
-      onSelectionChange={(key) => {
-        if (typeof key === "string") {
-          onChange(key);
+      onInputChange={onInputChange}
+      inputValue={fieldState.inputValue}
+      onSelectionChange={onSelectionChange}
+      width={"100%"}
+      allowsCustomValue
+      onBlur={() => {
+        if (fieldState.inputValue !== "") {
+          onChange(fieldState.inputValue);
+        }
+      }}
+      onKeyUp={(e) => {
+        if (e.key === "Enter") {
+          onChange(fieldState.inputValue);
         }
       }}
-      width={"100%"}
-      {...props}
-      defaultItems={items}
       menuTrigger="focus"
       container={container}
+      description={"Select a model or type one in and hit enter"}
+      defaultItems={items}
+      {...comboBoxProps}
     >
-      {(item) => (
-        <ComboBoxItem key={item.name} textValue={item.name}>
-          {item.name}
-        </ComboBoxItem>
-      )}
+      {(item) => {
+        return (
+          <ComboBoxItem key={item.id} textValue={item.name} id={item.id}>
+            {item.name}
+          </ComboBoxItem>
+        );
+      }}
     </ComboBox>
   );
 }
+
+export function ModelComboBox(props: ModelComboBoxProps) {
+  const [queryReference, loadQuery, disposeQuery] =
+    useQueryLoader<ModelComboBoxQuery>(modelsQuery);
+
+  useEffect(() => {
+    loadQuery({ providerKey: props.provider });
+    return () => disposeQuery();
+  }, [disposeQuery, loadQuery, props.provider]);
```

Comment:
What is the purpose of this vs lazyLoadQuery? Does it prevent triggering the nearest suspense boundary or something?